File: javascript/0021. Merge Two Sorted Lists.js
class Node {
  constructor(val, next = null) {
    this.val = val;
    this.next = next;
  }
}

const mergeTwoLists = (root1, root2) => {
  if (root1 === null) return root2;
  if (root2 === null) return root1;

  if (root1.val > root2.val) {
    const temp = root1;
    root1 = root2;
    root2 = temp;
  }

  let parentNode = root1;

  while (root2 && parentNode.next) {
    if (parentNode.next.val < root2.val) {
      parentNode = parentNode.next;
    } else {
      parentNode.next = new Node(root2.val, parentNode.next);
      parentNode = parentNode.next;
      root2 = root2.next;
    }
  }

  if (root2) {
    parentNode.next = root2;
  }

  return root1;
};

console.log(
  mergeTwoLists(
    {
      val: 1,
      next: { val: 2, next: { val: 4, next: null } },
    },
    {
      val: 1,
      next: { val: 3, next: { val: 4, next: null } },
    }
  )
);

File: python/0021. Merge Two Sorted Lists.py
# Definition for singly-linked list.
class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


class Solution:
    def mergeTwoLists(self, list_1: ListNode, list_2: ListNode) -> ListNode:
        if list_1 is None or list_2 is None:
            return list_1 or list_2

        node = prehead = ListNode()

        while list_1 is not None and list_2 is not None:
            if list_1.val > list_2.val:
                list_1, list_2 = list_2, list_1

            node.next = list_1
            node = list_1
            list_1 = list_1.next

        node.next = list_1 or list_2

        return prehead.next
